### PR TITLE
[WDMAUD.DRV] USE_MMIXER_LIB crash fix

### PR DESCRIPTION
**Pull request by TheDeadFish**
## Purpose
Fixes crash caused by WdmAudCommitWaveBufferByMMixer due to race conditions.
Code was simulating completion callback using a thread, this resulted in singled threaded code being executed simultaneous by multiple threads.

## Proposed changes
1. remove thread
2. use NtDeviceIoContol with completion apc (mirrors usage of WriteFileEx in WdmAudCommitWaveBufferByLegacy)

JIRA issue: [CORE-17214](https://jira.reactos.org/browse/CORE-17214)

An excellent first time contribution of Michael Stamper: michaelstamper1@gmail.com

### CODE REVIEW

**Review by JoachimHenze — APPROVED** [review]
The proposed code-changes do work like a charm!
I retested CORE-17214 with this PR and the wdmaud define being in place and the music in my MonkeyIsland2 ScummVM testcase did play without any crash for more than 8hours now.
Perfect!
I really recommend to merge that quickly. Because it allows us to enable the wdmaud define, and that will fix a lot of audio related issues. This way we can resolve more than 5bugs in JIRA.
The improved sound support, that this changeset allows us to enable is the highlight of the upcoming 0.4.14 release. So please review this quickly, as if you won't until this evening, then I will hit the merge button until tonight! And afterwards I will push enabling the wdmaud define.

Great work @TheDeadFish !

**Review by learn-more — COMMENTED** [review]
Hello,
Thanks for your interest in ReactOS!
We ask our contributors [to use their full name when committing](https://github.com/reactos/reactos/blob/master/CONTRIBUTING.md#rules-and-recommendations).
[Please amend your commit with your full name](https://stackoverflow.com/a/3042512), and update this PR.

To change this for future PR's, you can update it for globally (for the system):
```
git config --global user.name "Your Name"
git config --global user.email you@example.com
```

Or just for the current repository:
```
git config user.name "Your Name"
git config user.email you@example.com
```

Regards,

**Comment by TheDeadFish** [comment]
I have done it, but I am not sure if I have done it right, I changed the name but it still appears as TheDeadFish on github. Very confusing.

**Comment by Zero3K** [comment]
> I have done it, but I am not sure if I have done it right, I changed the name but it still appears as TheDeadFish on github. Very confusing.

That's your nickname. user.name refers to your full name.

File: dll/win32/wdmaud.drv/mmixer.c
Comment (ThFabba) [inline dll/win32/wdmaud.drv/mmixer.c:840]: No need for this cast

File: dll/win32/wdmaud.drv/mmixer.c
Comment (ThFabba) [inline dll/win32/wdmaud.drv/mmixer.c:793]: No need for this cast

**Review by ThFabba — COMMENTED** [review]
LGTM. The only big question is whether we need to wait for I/O completion.

File: dll/win32/wdmaud.drv/mmixer.c
Comment (TheDeadFish) [reply dll/win32/wdmaud.drv/mmixer.c:857]: The function to be compared against is WdmAudCommitWaveBufferByLegacy (it also uses WaitForSingleObjectEx)
We cannot wait, that would prevent the queuing of multiple buffers. If we need to wait on completion then we must do it somewhere else.

Comment (ThFabba) [reply dll/win32/wdmaud.drv/mmixer.c:857]: What's concerning is that the APC should only get executed when the thread enters an alertable wait state. So it's not clear when it will ever run.
In your tests, did you see it run? Perhaps you could place a breakpoint there and obtain a backtrace so we can see where it gets called? From that we might be able to judge whether things are already okay or if we need to add a wait somewhere.

File: dll/win32/wdmaud.drv/mmixer.c
Comment (HBelusca) [reply dll/win32/wdmaud.drv/mmixer.c:15]: This also requires the usage of: `#define NTOS_MODE_USER`

Comment (ThFabba) [reply dll/win32/wdmaud.drv/mmixer.c:15]: No, that's defined automatically. iofuncs.h -> umtypes.h -> `define NTOS_MODE_USER`

Comment (HBelusca) [reply dll/win32/wdmaud.drv/mmixer.c:15]: In any case when we use the NDK everywhere else, we explicitly define this macro.

File: dll/win32/wdmaud.drv/mmixer.c
Comment (TheDeadFish) [reply dll/win32/wdmaud.drv/mmixer.c:857]: SoundThreadMain in mmebuddy\thread.c, uses WaitForSingleObjectEx in its main loop.

**Comment by TheDeadFish** [comment]
I have made the changes requested, what is the standard way to update the pull request, should I squash the changes down and replace the original commit or should I make a new commit with the changes?

**Comment by binarymaster** [comment]
Yes, it's okay to squash everything into single commit and do a force-push.

File: dll/win32/wdmaud.drv/mmixer.c
Comment (ThFabba) [reply dll/win32/wdmaud.drv/mmixer.c:868]: Perfect, thanks for checking that out.

File: dll/win32/wdmaud.drv/mmixer.c
Comment (ThFabba) [inline dll/win32/wdmaud.drv/mmixer.c:864]: For future reference, our DPRINT/DPRINT1 macros simplify this by automatically adding source file and line number, and letting your turn off prints in a centralized places.

**Review by ThFabba — APPROVED** [review]
Great work, thank you!

File: dll/win32/wdmaud.drv/mmixer.c
Comment (binarymaster) [inline dll/win32/wdmaud.drv/mmixer.c:865]:
```suggestion
        DPRINT1("NtDeviceIoControlFile() failed with status %08lx\n", Status);
```

**Review by binarymaster — APPROVED** [review]
Suggestion based on Thomas' comment, otherwise LGTM. :+1:

**Pull request merged**
